runtime/tchannel/client.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/uber/tchannel-go/thrift"
+	"golang.org/x/net/context"
+)
+
+// client implements TChanClient and makes outgoing Thrift calls.
+type client struct {
+	ch          *tchannel.Channel
+	sc          *tchannel.SubChannel
+	serviceName string
+	opts        ClientOptions
+}
+
+// ClientOptions are options to customize the client.
+type ClientOptions struct {
+	// HostPort specifies a specific server to hit.
+	HostPort string
+}
+
+// NewClient returns a Client that makes calls over the given tchannel to the given Hyperbahn service.
+func NewClient(ch *tchannel.Channel, serviceName string, opts *ClientOptions) TChanClient {
+	client := &client{
+		ch:          ch,
+		sc:          ch.GetSubChannel(serviceName),
+		serviceName: serviceName,
+	}
+	if opts != nil {
+		client.opts = *opts
+	}
+	return client
+}
+
+func (c *client) startCall(ctx context.Context, method string, callOptions *tchannel.CallOptions) (*tchannel.OutboundCall, error) {
+	if c.opts.HostPort != "" {
+		return c.ch.BeginCall(ctx, c.opts.HostPort, c.serviceName, method, callOptions)
+	}
+	return c.sc.BeginCall(ctx, method, callOptions)
+}
+
+func writeArgs(call *tchannel.OutboundCall, headers map[string]string, req RWTStruct) error {
+	writer, err := call.Arg2Writer()
+	if err != nil {
+		return err
+	}
+	headers = tchannel.InjectOutboundSpan(call.Response(), headers)
+	if err := thrift.WriteHeaders(writer, headers); err != nil {
+		return err
+	}
+	if err := writer.Close(); err != nil {
+		return err
+	}
+
+	writer, err = call.Arg3Writer()
+	if err != nil {
+		return err
+	}
+
+	if err := WriteStruct(writer, req); err != nil {
+		return err
+	}
+
+	return writer.Close()
+}
+
+// readResponse reads the response struct into resp, and returns:
+// (response headers, whether there was an application error, unexpected error).
+func readResponse(response *tchannel.OutboundCallResponse, resp RWTStruct) (map[string]string, bool, error) {
+	reader, err := response.Arg2Reader()
+	if err != nil {
+		return nil, false, err
+	}
+
+	headers, err := thrift.ReadHeaders(reader)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if err := EnsureEmpty(reader, "reading response headers"); err != nil {
+		return nil, false, err
+	}
+
+	if err := reader.Close(); err != nil {
+		return nil, false, err
+	}
+
+	success := !response.ApplicationError()
+	reader, err = response.Arg3Reader()
+	if err != nil {
+		return headers, success, err
+	}
+
+	if err := ReadStruct(reader, resp); err != nil {
+		return headers, success, err
+	}
+
+	if err := EnsureEmpty(reader, "reading response body"); err != nil {
+		return nil, false, err
+	}
+
+	return headers, success, reader.Close()
+}
+
+func (c *client) Call(ctx thrift.Context, thriftService, methodName string, req, resp RWTStruct) (bool, error) {
+	var (
+		headers = ctx.Headers()
+
+		respHeaders map[string]string
+		isOK        bool
+	)
+
+	err := c.ch.RunWithRetry(ctx, func(ctx context.Context, rs *tchannel.RequestState) error {
```

Comment:
Talking to @jcorbin he mentioned that closures are really fast except that they "close" over everything, aka allocating a closure in a big method will move a lot of variables to the heap instead of the stack.